hibike/hibike_simulator.py

```diff
@@ -1,133 +1,156 @@
+    while True:
+        instruction, args = pipeFromChild.recv()
+        if instruction == "enumerate_all":
+            for device in devices:
+                device.instruction_queue.put(("ping", []))
+        elif instruction == "subscribe_device":
+            uid = args[0]
+            if uid in uid_to_queue:
+                uid_to_queue[uid].put(("subscribe", args))
+            else:
+                print('not found uid')
+
+def runDeviceWrite(instruction_queue, fake_device_queue):
+    # instruction_queue is input
+    # fake_device_queue is output to smart sensor
+    while True:
+        instruction, args = instruction_queue.get()
+
+        if instruction == "ping":
+            fake_device_queue.put(0)
+        elif instruction == "subscribe":
+            uid, delay, params = tuple(args)
+            fake_device_queue.put(delay)
+            fake_device_queue.put(params)
+
+def runDeviceRead(fake_device_readings, stateQueue, instruction_queue):
+    while True:
+        result, values = fake_device_readings.get()
+        res = None
+        if result == "device_values":
+            res = [result, values]
+        elif result == "device_subscribed":
+            uid = values[0]
+            if uid in uid_to_queue:
+                del uid_to_queue[uid]
+            uid_to_queue[uid] = instruction_queue
+            res = ["device_subscribed", values]
         else:
-            return low + (random.random() - .5) * noise
-
-############
-## MOTORS ##
-############
-
-    def readValue(self, UID, param):
-        #param values: 0 - delay, 1 - last_time, 2 - data
+            print("Some error code received")
+            print(result, values)
+
+        if res is not None:
+            stateQueue.put(res)
+
+def runFakeSubscription(uid, fake_device_queue, fake_device_readings):
+    # fake_device_queue is input commands
+    # fake_device_readings is output of fake readings
+    timeout = None
+    new_delay = False
+    while True:
         try:
-            return self.subscribedTo[UID][param]
-        except:
-            return 1
-
-    def writeValue(self, UID, param, value):
-        if param >= len(self.subscribedTo[UID]):
-            return 1
-        self.subscribedTo[UID][param] = value
-        return 0
+            delay = fake_device_queue.get(block=True, timeout=timeout)
+            new_delay = True
+        except queue.Empty:
+            pass
+        # Simulate enumeration
+        if delay == 0:
+            params = list(paramMap[uid_to_device_id(uid)].keys())
+            fake_device_readings.put(("device_subscribed", [uid, 0, params]))
+        # Simulate subscription
+        elif delay > 0:
+            if new_delay:
+                params = fake_device_queue.get()
+            param_types = [paramMap[uid_to_device_id(uid)][param][1] for param in params]
+            params_and_values = {}
+            if __name__ == "__main__":
```

Comment:
What is going on here??? Even a global var like `DETERMINISTIC_MODE` would be less hacky.

(Something else to consider is seeding the RNG at main startup, instead of having a branch here)